hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/producers/StreamFileProducerConcurrent.java

```diff
@@ -0,0 +1,273 @@
+    private CompletableFuture<Bytes> lastRecordHashingResultNMinus1 = null;
+    /** Future for running hash results of previous, previous lastRecordHashingResult */
+    private CompletableFuture<Bytes> lastRecordHashingResultNMinus2 = null;
+    /** Future for running hash results of previous, previous, previous lastRecordHashingResult */
+    private CompletableFuture<Bytes> lastRecordHashingResultNMinus3 = null;
+    /**
+     * Future for the current record file writer, this will always get you the record file writer for the current file
+     * being written
+     */
+    private CompletableFuture<BlockRecordWriter> currentRecordFileWriter = null;
+    /** Set in {@link #switchBlocks(long, long, Instant)}, keeps track of the current block number. */
+    private long currentBlockNumber;
+
+    /**
+     * Construct {@link StreamFileProducerConcurrent}
+     *
+     * @param nodeInfo the current node information
+     * @param format The format to use for the record stream
+     * @param writerFactory the factory used to create new {@link BlockRecordWriter} instances
+     * @param executorService The executor service to use for background threads
+     */
+    @Inject
+    public StreamFileProducerConcurrent(
+            @NonNull final SelfNodeInfo nodeInfo,
+            @NonNull final BlockRecordFormat format,
+            @NonNull final BlockRecordWriterFactory writerFactory,
+            @CommonExecutor @NonNull final ExecutorService executorService) {
+        this.writerFactory = requireNonNull(writerFactory);
+        this.format = requireNonNull(format);
+        hapiVersion = nodeInfo.hapiVersion();
+        this.executorService = requireNonNull(executorService);
+    }
+
+    // =================================================================================================================
+    // Implementation of BlockRecordStreamProducer methods
+    //
+    // These methods are ALWAYS called on the "handle" thread.
+
+    /** {@inheritDoc} */
+    @Override
+    public void initRunningHash(@NonNull final RunningHashes runningHashes) {
+        if (lastRecordHashingResult != null) {
+            throw new IllegalStateException("initRunningHash() can only be called once");
+        }
+
+        if (runningHashes.runningHash() == null) {
+            throw new IllegalArgumentException("The initial running hash cannot be null");
+        }
+
+        lastRecordHashingResult = completedFuture(runningHashes.runningHash());
+        lastRecordHashingResultNMinus1 = completedFuture(runningHashes.nMinus1RunningHash());
+        lastRecordHashingResultNMinus2 = completedFuture(runningHashes.nMinus2RunningHash());
+        lastRecordHashingResultNMinus3 = completedFuture(runningHashes.nMinus3RunningHash());
+    }
+
+    /** {@inheritDoc} */
+    @NonNull
+    @Override
+    public Bytes getRunningHash() {
+        assert lastRecordHashingResult != null : "initRunningHash() must be called before getRunningHash()";
```

Comment:
I agree in a public API that would be preferable. In this case, such a check is just pure overhead. It isn't possible to get this wrong just some of the time -- it is either always right, or always wrong. In fact, the method will throw even if this assert isn't there. The only reason for having it, is to make it clear when a test fails (since -ea is only used in testing) what happened and how to debug. It could be a comment. But I don't know that it is justified to be a full check every time this method is called.

Reply:
I am not convinced. The method is public, which means somebody may call it from another context in the future. Even if it is unlikely, using an explicit check is a no-brainer. If an `assert` is used, I have to consider carefully if it is justified and reconsider each time I change something. But anyway, not critical and should not hold back the PR.